forem/forem
# Create mention autocomplete: initial storybook component

## What type of PR is this? (check all applicable)


## Description

This PR is the first piece of work towards adding mention autocomplete functionality. It introduces the base Crayons component, but for now is only used in Storybook with a completely mocked dataset.

A subsequent PR will cover implementing this component in the real app with the real user search network call.

## Related Tickets & Documents

[Create mention autocomplete](https://github.com/forem/rfcs/pull/16)

## QA Instructions, Screenshots, Recordings

In the storybook text area, start typing. When you type '@' plus two characters, you should start seeing results in the dropdown, appropriately positioned depending on where your cursor was when you started the search.

The 'search' functionality should exit when you either:
- Select an option with a click
- Select an option with keyboard Enter press
- Press 'Space' to accept what you've currently typed instead of a suggestion

https://user-images.githubusercontent.com/20773163/109792439-6e2f2500-7c0b-11eb-8673-3a64c054bda3.mp4


Also some screenshots of the theme variants:

![screenshot showing the night, pink and hacker themes](https://user-images.githubusercontent.com/20773163/109508792-22a83a00-7a98-11eb-8b14-bfd7a513e82a.png)


### UI accessibility concerns?

I've gone through the [W3 WAI-ARIA guidance on comboboxes](https://www.w3.org/TR/wai-aria-practices-1.2/#combobox) and the Storybook component here appears to mirror the properties and functionality in their [Autocomplete example](https://www.w3.org/TR/wai-aria-practices-1.2/examples/combobox/combobox-autocomplete-list.html).

I've tested the component in Safari with VoiceOver, and again it behaves the same as the W3 guidance/examples. The input is appropriately announced as a combobox, and the options can be scrolled through with the up and down arrows. Escape closes the dropdown, as does tabbing away.

**Short recording of VoiceOver output:**


https://user-images.githubusercontent.com/20773163/109792711-b5b5b100-7c0b-11eb-8d2c-e682a5ab5ce1.mp4


Axe flags a couple of issues but I believe these are false positives:

- 1 complaint that the input doesn't have a `listbox` as a descendant
- 1 complaint that the `listbox` doesn't have an accessible name

I think these are false positives as the Reach Combox is using a portal to display the listbox. It's linked by ID to the input, and from manual testing & inspection in Firefox/Chrome/Safari the accessible names are being surfaced fine.

## Added tests?

- [X] Yes - eventually we will also have end to end tests but for the moment I have added component level tests
      have not been included_

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

- [X] I've updated the [Storybook](https://storybook.forem.com/) (for Crayons components)
      or in a [forem.dev](http://forem.dev) post
      replace this line with details on why this change doesn't need to be
      shared_

## [optional] Are there any post deployment tasks we need to perform?

N/A - no user facing changes yet

## [optional] What gif best describes this PR or how it makes you feel?

![the coneheads say it's also visually pleasing](https://i.giphy.com/media/d31xcQedqXbduoyA/giphy.webp)
